File: ungs-docsys-ms/src/main/java/com/ungs/docsys/controllers/JobApplicationController.java
package com.ungs.docsys.controllers;

import com.ungs.docsys.dtos.JobApplicationRequestDto;
import com.ungs.docsys.dtos.JobApplicationResponseDto;
import com.ungs.docsys.security.JwtUtil;
import com.ungs.docsys.services.JobApplicationService;
import com.ungs.docsys.utils.ExcelExportUtils;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import jakarta.validation.Valid;
import lombok.AllArgsConstructor;
import org.springframework.http.*;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.io.IOException;
import java.util.List;

@RestController
@RequestMapping("/v1/job-applications")
@Validated
@AllArgsConstructor
public class JobApplicationController {

    private final JobApplicationService jobApplicationService;
    private final JwtUtil jwtUtil;

    @Operation(summary = "Create a job application", description = "Creates a new job application for the authenticated user.")
    @ApiResponse(responseCode = "200", description = "Job application created successfully")
    @ApiResponse(responseCode = "400", description = "Invalid request data")
    @ApiResponse(responseCode = "500", description = "Internal server error")
    @PostMapping()
    public ResponseEntity<JobApplicationResponseDto> create(
            @RequestHeader("Authorization") String authorization,
            @Valid @RequestBody JobApplicationRequestDto request) {
        return ResponseEntity.ok(jobApplicationService.create(request, jwtUtil.extractUserClaim(authorization)));
    }

    @Operation(summary = "Delete a job application")
    @ApiResponse(responseCode = "204", description = "Job application deleted successfully")
    @ApiResponse(responseCode = "404", description = "Job application not found")
    @ApiResponse(responseCode = "500", description = "Internal server error")
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> delete(@PathVariable Long id) {
        return jobApplicationService.delete(id)
                ? ResponseEntity.noContent().build()
                : ResponseEntity.status(HttpStatus.NOT_MODIFIED).build();
    }

    @Operation(summary = "Update a job application")
    @ApiResponse(responseCode = "200", description = "Job application updated successfully")
    @ApiResponse(responseCode = "404", description = "Job application not found")
    @ApiResponse(responseCode = "403", description = "Forbidden - user not authorized to update this job application")
    @ApiResponse(responseCode = "500", description = "Internal server error")
    @PatchMapping("/{id}")
    public ResponseEntity<JobApplicationResponseDto> update(
            @PathVariable Long id,
            @RequestBody JobApplicationRequestDto request) {
        JobApplicationResponseDto updated = jobApplicationService.partiallyUpdate(id, request);
        return ResponseEntity.ok(updated);
    }
    @Operation(summary = "Get all jobs applications")
    @ApiResponse(responseCode = "200", description = "Job application updated successfully")
    @ApiResponse(responseCode = "404", description = "Job application not found")
    @ApiResponse(responseCode = "403", description = "Forbidden - user not authorized to update this job application")
    @ApiResponse(responseCode = "500", description = "Internal server error")
    @GetMapping()
    public ResponseEntity<List<JobApplicationResponseDto>> getAll() {
        return ResponseEntity.ok(jobApplicationService.getAll());
    }

    @Operation(summary = "Get identification type by id")
    @ApiResponse(responseCode = "200", description = "Success")
    @ApiResponse(responseCode = "404", description = "Not found")
    @GetMapping("/{id}")
    public ResponseEntity<JobApplicationResponseDto> getById(@PathVariable Long id) {
        return ResponseEntity.ok(jobApplicationService.getById(id));
    }

    @Operation(summary = "Export a job application")
    @ApiResponse(responseCode = "200", description = "Job application exported successfully")
    @ApiResponse(responseCode = "404", description = "Job application not found")
    @GetMapping("/{id}/export")
    public ResponseEntity<byte[]> export(
            @RequestHeader("Authorization") String authorization,
            @PathVariable Long id) throws IOException {
        JobApplicationResponseDto jobApplicationResponseDto = jobApplicationService.getById(id);
        byte[] data = jobApplicationService.exportToExcel(id, jwtUtil.extractUserClaim(authorization));
        String fileName = ExcelExportUtils.generateExportFileName(jobApplicationResponseDto.getTitle());
        return ResponseEntity.ok()
                .contentType(MediaType.APPLICATION_OCTET_STREAM)
                .header(HttpHeaders.CONTENT_DISPOSITION,
                        ContentDisposition.attachment().filename(fileName).build().toString())
                .body(data);
    }
}

File: ungs-docsys-bff/src/modules/job-applications/controllers/job-application.controller.ts
import { Controller, Post, Patch, Body, Param, Headers, HttpCode, HttpStatus, UnauthorizedException, HttpException, Delete, Get, Res} from '@nestjs/common';
import { JobApplicationService } from '../services/job-application.service';
import { JobApplicationRequestDto } from '../dtos/job-application-request.dto';
import { JobApplicationUpdateRequestDto } from '../dtos/job-application-update-request.dts';
import { JobApplicationResponseDto } from '../dtos/job-application-response.dto';
import { ApiOperation, ApiResponse } from '@nestjs/swagger';
import { Response } from 'express';

@Controller("/v1/job-applications")
export class JobApplicationController {
    constructor(private readonly jobApplicationService: JobApplicationService) { }

    @Post()
    @ApiOperation({ summary: 'Create Job Application' })
    @ApiResponse({
        status: 200,
        description: 'Successfully created job application',
        type: JobApplicationResponseDto
    })
    async create(@Body() request: JobApplicationRequestDto, @Headers('authorization') authHeader: string): Promise<JobApplicationResponseDto> {
        return this.jobApplicationService.create(request, authHeader);
    }

    @Delete(':id')
    @HttpCode(HttpStatus.NO_CONTENT)
    @ApiOperation({ summary: 'Delete Job Application by ID' })
    @ApiResponse({
        status: 204,
        description: 'Successfully deleted (soft delete)'
    })
    @ApiResponse({
        status: 304,
        description: 'Not modified'
    })
    async delete(@Param('id') id: number, @Headers('authorization') authHeader: string): Promise<void> {
        const token = this.extractToken(authHeader);
        const deleted = await this.jobApplicationService.delete(id, token);

        if (!deleted) {
            throw new HttpException('Not modified', HttpStatus.NOT_MODIFIED);
        }
    }

    @Patch(':id')
    @ApiOperation({ summary: 'Update Job Application by ID' })
    @ApiResponse({
        status: 200,
        description: 'Successfully updated job application',
        type: JobApplicationResponseDto
    })
    async update(
        @Param('id') id: number, @Body() request: JobApplicationUpdateRequestDto, @Headers('authorization') authHeader: string): Promise<JobApplicationResponseDto> {
        const token = authHeader?.replace('Bearer ', '');
        return this.jobApplicationService.update(id, request, token);
    }

    @Get()
    @ApiOperation({ summary: 'Get all Requirement types' })
    @ApiResponse({
        status: 200,
        description: 'Requirement type details',
        type: [JobApplicationResponseDto],
    })
    async getAll(@Headers('authorization') authorization: string,): Promise<JobApplicationResponseDto[]> {
        return await this.jobApplicationService.getAll(authorization);
    }

    private extractToken(authHeader: string): string {
        if (!authHeader || !authHeader.startsWith('Bearer ')) {
            throw new UnauthorizedException('Authorization token is missing or malformed');
        }
        return authHeader.replace('Bearer ', '');
    }

    @Get(':id')
      @ApiOperation({ summary: 'Get Job application by id' })
      @ApiResponse({
        status: 200,
        description: 'Job application details',
        type: JobApplicationResponseDto,
      })
      async getById(@Headers('authorization') authorization: string, @Param('id') id: number): Promise<JobApplicationResponseDto> {
        return await this.jobApplicationService.getById(id, authorization);
      }

    @Get(':id/export')
    @ApiOperation({ summary: 'Export job application to Excel' })
    @ApiResponse({ status: 200, description: 'Excel file downloaded' })
    async export(
        @Headers('authorization') authorization: string, 
        @Param('id') id: number, 
        @Res() res: Response): Promise<void> {
        const { data, headers } = await this.jobApplicationService.export(id, authorization);

        res.set({
            'Content-Type': headers['content-type'] || 'application/octet-stream',
            'Content-Disposition': headers['content-disposition'],
        });

        res.send(data);
    }
}